src/main/java/org/apache/pulsar/io/gcp/PubsubSource.java

```diff
@@ -52,16 +49,16 @@ public void open(Map<String, Object> config, SourceContext sourceContext) throws
             .build();
 
     Subscriber.Builder subscriberBuilder = Subscriber.newBuilder(projectSubscriptionName,
-        (PubsubMessage message, AckReplyConsumer consumer) -> {
-          Record<byte[]> record = new PubsubRecord(sourceContext.getOutputTopic(), message);
-          try {
-            queue.put(record);
-            consumer.ack();
-          } catch (InterruptedException e) {
-            consumer.nack();
-            throw new RuntimeException(e);
-          }
-        })
+            (PubsubMessage message, AckReplyConsumer consumer) -> {
+              Record<byte[]> record = new PubsubRecord(sourceContext.getOutputTopic(), message);
+              try {
+                consume(record);
```

Comment:
It looks like this just moves the LinkedBlockingQueue implementation into PushSource: https://github.com/apache/pulsar/blob/426ad3e53ba621c442bce89eb208add8f9d1563e/pulsar-io/core/src/main/java/org/apache/pulsar/io/core/PushSource.java#L72. It looks like the right thing to do here is to override the ack method in Record and then call Pub/Sub's ack in there: https://github.com/apache/pulsar/blob/426ad3e53ba621c442bce89eb208add8f9d1563e/pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/Record.java#L109.